sonic_data_client/events_client.go

```diff
@@ -48,48 +87,192 @@ func NewEventClient(paths []*gnmipb.Path, prefix *gnmipb.Path, logLevel int) (Cl
+                    log.V(7).Infof("evtc.heartbeat_interval is set to %d", val)
+                    set_heartbeat(val)
+                }
+            }
+        }
     }
-    js := string(j)
-    evtc.subs_handle = C.events_init_subscriber_wrap(C.CString(js))
+
+    C.swssSetLogPriority(C.int(logLevel))
+
+    /* Init subscriber with cache use and defined time out */
+    evtc.subs_handle = C.events_init_subscriber_wrap(true, C.int(SUBSCRIBER_TIMEOUT))
     evtc.stopped = 0
 
+    /* Init list & counters */
+    evtc.counters = make(map[string]uint64)
+
+    for _, key := range STATS_CUMULATIVE_KEYS {
+        evtc.counters[key] = 0
+    }
+
+    for _, key := range STATS_ABSOLUTE_KEYS {
+        evtc.counters[key] = 0
+    }
+
+    for i := 0; i < len(evtc.last_latencies); i++ {
+        evtc.last_latencies[i] = 0
+    }
+    evtc.last_latency_index = 0
+    evtc.last_errors = 0
+    evtc.last_latency_full = false
+
     log.V(7).Infof("NewEventClient constructed. logLevel=%d", logLevel)
 
     return &evtc, nil
 }
 
+
+func compute_latency(evtc *EventClient) {
+    if evtc.last_latency_full {
+        var total uint64 = 0
+        var cnt uint64 = 0
+
+        for _, v := range evtc.last_latencies {
+            if v > 0 {
+                total += v
+                cnt += 1
+            }
+        }
+        evtc.counters[LATENCY] = (uint64) (total/cnt/1000/1000)
+    }
+}
+
+func update_stats(evtc *EventClient) {
+    /* Wait for any update */
+    db_counters := make(map[string]uint64)
+    var wr_counters *map[string]uint64 = nil
+    var rclient *redis.Client
+
+    for evtc.stopped == 0 {
+        var val uint64
+
+        compute_latency(evtc)
+        for _, val = range evtc.counters {
+            if val != 0 {
+                break
+            }
+        }
+        if val != 0 {
+            break
+        }
+        time.Sleep(time.Second)
+    }
+    
+    if evtc.stopped == 0 {
+        ns := sdcfg.GetDbDefaultNamespace()
+
+        rclient = redis.NewClient(&redis.Options{
+            Network:    "tcp",
+            Addr:       sdcfg.GetDbTcpAddr("COUNTERS_DB", ns),
+            Password:   "", // no password set,
+            DB:         sdcfg.GetDbId("COUNTERS_DB", ns),
+            DialTimeout:0,
+        })
+
+
+        // Init current values for cumulative keys and clear for absolute 
+        for _, key := range STATS_CUMULATIVE_KEYS {
+            fv, err := rclient.HGetAll(key).Result()
+            if err != nil {
+                number, errC := strconv.ParseUint(fv[STATS_FIELD_NAME], 10, 64)
+                if errC == nil {
+                    db_counters[key] = number
+                }
+            }
+        }
+        for _, key := range STATS_ABSOLUTE_KEYS {
+            db_counters[key] = 0
+        }
+    }
+
+    for evtc.stopped == 0 {
+        tmp_counters := make(map[string]uint64)
+
+        // compute latency
+        if evtc.last_latency_full {
```

Comment:
When are you setting back to false?

Reply:
Why are we checking twice, we call compute latency on 157?

Reply:
Set to False in NewEventClient.

Missed to replace the code with the function.